File: combigrid/examples/c++/global_interpolation/sin_func_full_sg_interpolation.cpp
#include <cmath>
#include <iostream>
#include <ostream>
#include <sgpp/base/datatypes/DataVector.hpp>
#include <sgpp/combigrid/functions/level_to_grid_point_count_functions/level_to_grid_point_count_functions.hpp>
#include <sgpp/combigrid/functions/node_generation_functions/getter/clenshaw_curtis_node_generation_function_getter.hpp>
#include <sgpp/combigrid/functions/source_functions/source_function.hpp>
#include <sgpp/combigrid/grids/sparse_grid.hpp>
#include <sgpp/combigrid/operators/interpolation/interpolation.hpp>
#include <sgpp/combigrid/sparse_grid_generation_instructions/full_sg_gen_instruction.hpp>

using namespace sgpp::combigrid;
using DataVector = sgpp::base::DataVector;

double func(const DataVector& point) {
  double result = 1;

  for (size_t dim = 0; dim < point.size(); dim++) {
    result *= std::cos(4 * M_PI * point[dim]);
  }

  return result;
}

int main() {
  const SourceFunc sourceFunc(func);

  FullSGGenInstr genInstr(10, 3);
  genInstr.setBounds({0, 1});
  genInstr.setNodeGenFunc(getClenshawCurtisNodeGenFunc());
  genInstr.setLvl2GPCntFunc(doublingLvl2GPCntFunction);
  genInstr.setBoundaryIndexOffset(0);

  const SparseGrid sg(genInstr);

  const DataVector point{0.9173, 0.13494, 0.4184};
  const double result = interpolate(sourceFunc, point, sg);

  std::cout << "Result: " << result << std::endl;
  std::cout << "Expected: " << func(point) << std::endl;

  return 0;
}

File: combigrid/examples/python/sin_func_full_sg_interpolation.py
#!/usr/bin/python
# Copyright (C) 2008-today The SG++ project
# This file is part of the SG++ project. For conditions of distribution and
# use, please see the copyright notice provided with SG++ or at
# sgpp.sparsegrids.org

## \page example_combigrid_sin_func_full_sg_interpolation_py Global interpolation on a complete sparse grid (Python)
##
## Python version of combigrid/examples/c++/global_interpolation/sin_func_full_sg_interpolation.cpp.
##
## The function
## \f[
##   f\colon [0, 1]^3 \to \mathbb{R},\quad
##   f(x) := \prod_{k=0}^{2} \cos(4 \pi x_k)
## \f]
## is interpolated with the combination technique on a complete sparse grid of maximum level 10,
## using Clenshaw-Curtis nodes with doubling growth, and compared to the exact function value.

import math
import pysgpp


## The source function receives the coordinates of a point as a tuple of floats.
def func(point):
  result = 1.0

  for dim in range(len(point)):
    result *= math.cos(4 * math.pi * point[dim])

  return result


sourceFunc = pysgpp.SourceFunc(func)

## Configure the generation instruction: all level multi-indices with 1-norm at most 10 in three
## dimensions, Clenshaw-Curtis nodes, doubling growth, and boundary nodes on every level.
genInstr = pysgpp.CompleteSGGenInstr(10, 3)
genInstr.setDomain((0, 1))
genInstr.setNodeGenFunc(pysgpp.getClenshawCurtisNodeGenFunc())
genInstr.setLvl2GPCntFunc(pysgpp.doublingLvl2GPCntFunction)
genInstr.setBoundaryLevelOffset(0)

## Generate the sparse grid, i.e., all tensor grids with a nonzero combination coefficient.
sg = pysgpp.SparseGrid(genInstr)

point = pysgpp.DataVector([0.9173, 0.13494, 0.4184])
result = pysgpp.interpolate(sourceFunc, point, sg)

print(f"Result: {result}")
print(f"Expected: {func(point)}")
